lib/ABI/AArch64/AAPCS64.h

```diff
@@ -0,0 +1,104 @@
+#pragma once
+
+//
+// This file is distributed under the MIT Licence. See LICENSE.md for details.
+//
+
+#include "llvm/ADT/StringRef.h"
+
+#include "revng/ABI/Definition.h"
+
+namespace abi::predefined::AArch64::AAPCS64 {
+
+constexpr llvm::StringRef Serialized = R"(---
+ABI: AAPCS64
+
+ArgumentsArePositionBased: false
+OnlyStartDoubleArgumentsFromAnEvenRegister: true
+ArgumentsCanBeSplitBetweenRegistersAndStack: true
+UsePointerToCopyForStackArguments: false
+CalleeIsResponsibleForStackCleanup: false
+
+StackAlignment: 16
+
+MaximumGPRsPerAggregateArgument: 2
+MaximumGPRsPerAggregateReturnValue: 2
+MaximumGPRsPerScalarArgument: 2
+MaximumGPRsPerScalarReturnValue: 2
+
+GeneralPurposeArgumentRegisters:
+  - x0_aarch64
+  - x1_aarch64
+  - x2_aarch64
+  - x3_aarch64
+  - x4_aarch64
+  - x5_aarch64
+  - x6_aarch64
+  - x7_aarch64
+GeneralPurposeReturnValueRegisters:
+  - x0_aarch64
+  - x1_aarch64
+  - x2_aarch64
+  - x3_aarch64
+  - x4_aarch64
+  - x5_aarch64
+  - x6_aarch64
+  - x7_aarch64
+VectorArgumentRegisters:
+  - v0_aarch64
+  - v1_aarch64
+  - v2_aarch64
+  - v3_aarch64
+  - v4_aarch64
+  - v5_aarch64
+  - v6_aarch64
+  - v7_aarch64
+VectorReturnValueRegisters:
+  - v0_aarch64
+  - v1_aarch64
+  - v2_aarch64
+  - v3_aarch64
+  - v4_aarch64
+  - v5_aarch64
+  - v6_aarch64
+  - v7_aarch64
+CalleeSavedRegisters:
+  - x19_aarch64
+  - x20_aarch64
+  - x21_aarch64
+  - x22_aarch64
+  - x23_aarch64
+  - x24_aarch64
+  - x25_aarch64
+  - x26_aarch64
+  - x27_aarch64
+  - x28_aarch64
+  - x29_aarch64
+  - sp_aarch64
+  - v8_aarch64
+  - v9_aarch64
+  - v10_aarch64
+  - v11_aarch64
+  - v12_aarch64
+  - v13_aarch64
+  - v14_aarch64
+  - v15_aarch64
+
+ReturnValueLocationRegister: x8_aarch64
+
+Types:
+  - Size: 1
+    AlignAt: 1
+  - Size: 2
+    AlignAt: 2
+  - Size: 4
+    AlignAt: 4
+  - Size: 8
+    AlignAt: 8
+  - Size: 16
+    AlignAt: 16
```

Comment:
I don't think that's possible because then we wouldn't have a clear way of telling if alignment should be defined for given size, e.g. 3. We could hard-limit it to power-of-twos, BUT consider that this data-structure will be reused for the specifics of vector-register placement as well, so these members will have more fields once we start supporting those.

P. S. On the second thought, I might be fine with something like
```yaml
Types:
  - Size: 1
  - Size: 2
  - Size: 4
  - Size: 8
  - Size: 16
```

What do you think?

Reply:
I'm not sure how should I interpret this. What does it mean if `- Size: 2`, say, is not there?

Reply:
That there is no 16-bit integer support, I'd say.

Reply:
As in `int16_t Array[100]` is impossible and would actually look like `int32_t Array[100]` in memory.

As in, nothing can be aligned at double-byte boundary.

Reply:
I guess this struct would be a better example:
```cpp
struct Whatever {
  uint16_t a;
  uint16_t b;
};
```
-----------
When alignment at double-byte boundary is available, it's:
```cpp
struct Whatever {
  uint16_t a; // aligned at 2
  uint16_t b; // aligned at 2
}; // sizeof(Whatever) == 4;
```
When it isn't:
```cpp
struct Whatever {
  uint16_t a; // aligned at 4
  uint16_t b; // aligned at 4
}; // sizeof(Whatever) == 8;
```

Reply:
Can we replace `Types` with `PossibleAlignments` that is a list of integers and implicitly say that each type is aligned to its size or to the next possible alignment?

`Types` doesn't really reflect what we're trying to express here well.

Reply:
It sure doesn't. The main reason for representing the data this way is to anticipate for a bunch of vector-register related data we'll have to stuff into the same poor ~~turkey~~ definition object.

But beside that, just a list does not cut it in cases where alignment is not the same as size, for example here: https://github.com/Cvelth/revng/blob/feature/stack-argument-alignment/share/revng/abi/SystemZ_s390x.yml#L77-L78.

Reply:
Ah jeez. Can we bail out/abort in case a scalar type with size not mentioned here?
I'm confused about what happens if a type *is not* here.

Reply:
> I'm confused about what happens if a type is *not* here.

It's simple, it just gets extended to the next one.

For example, if we have a 12-byte scalar, it would just get extended to 16 bytes (if ABI supports those, i.e. there is a 16-byte entry in this table, assert otherwise) and aligned the same way 16-byte scalars are aligned.

Reply:
Consider that when I say "we have no floating point support", I mean we have no "vector register" support. If those floats are passed though the stack, they should work fine.

I won't lie and say that it's exhaustively tested, but it's not totally untested either.

Reply:
I'd prefer to call this `ScalarTypes` and fail if any scalar is not present here. Much simpler and the there aren't many scalar types anyway.

Reply:
Consider that this struct is the basis for calculation alignment for any arbitrary type in the model. And IMO, we should be able to figure it any for any valid *sized* type (as long as it is not self-referencing or something broken like that. If `size()` works, `ABI.alignment()` should too).

The alignment of aggregates is, luckily, super simple - it's the maximum of the alignments of its members.

Now, the question, where do you draw the border of what is a scalar in this context?

P. S. Also, it feels like the discussion suddenly changed to only affect the naming, should I see that as a confirmation that the internals of `PrimitiveTypeConstraints` are fine as they are?

P. P. S. We should probably also rename it to `ScalarTypeConstraints` while we're at it. Thoughts?

Reply:
> we should be able to figure it any for any valid sized type

Yes, but given that the set of scalar/primitive types conceived by an ABI is finite and that all larger types are aggregate and their alignment can be (or actually, is) computed not looking at this list, but simply at their fields, I'd say it's simpler to call this either `ScalarTypes` or `PrimitiveTypes` and ban any scalar/primitive not in there.

Reply:
What do you think about it now?

See `ScalarType.h` header, the last two members of the TupleTree generated from `Definition.h` and the ABI yamls.

Reply:
So we now ban all integers that are not listed here? If so, OK, feel free to mark it solved!

Reply:
Yep, and `AlignsAt` is only listed for the primitives whose size is different from the alignment.